src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/PolymorphicTypeResolver.cs

```diff
@@ -0,0 +1,252 @@
+        private readonly Dictionary<string, DerivedJsonTypeInfo>? _discriminatorIdtoType;
+
+        public PolymorphicTypeResolver(JsonConverter baseConverter, IJsonPolymorphicTypeConfiguration configuration, JsonSerializerOptions options)
+        {
+            _options = options;
+            BaseType = configuration.BaseType;
+            UnknownDerivedTypeHandling = configuration.UnknownDerivedTypeHandling;
+            IgnoreUnrecognizedTypeDiscriminators = configuration.IgnoreUnrecognizedTypeDiscriminators;
+
+            if (!IsSupportedPolymorphicBaseType(BaseType))
+            {
+                ThrowHelper.ThrowInvalidOperationException_TypeDoesNotSupportPolymorphism(BaseType);
+            }
+
+            bool containsDerivedTypes = false;
+            foreach ((Type derivedType, string? typeDiscriminatorId) in configuration.GetSupportedDerivedTypes())
+            {
+                if (!IsSupportedDerivedType(BaseType, derivedType) ||
+                    (derivedType.IsAbstract && UnknownDerivedTypeHandling != JsonUnknownDerivedTypeHandling.FallbackToNearestAncestor))
+                {
+                    ThrowHelper.ThrowInvalidOperationException_DerivedTypeNotSupported(BaseType, derivedType);
+                }
+
+                if (typeDiscriminatorId is not null)
+                {
+                    UsesTypeDiscriminators = true;
+                }
+
+                var derivedJsonTypeInfo = new DerivedJsonTypeInfo(derivedType, typeDiscriminatorId);
+
+                if (!_typeToDiscriminatorId.TryAdd(derivedType, derivedJsonTypeInfo))
+                {
+                    ThrowHelper.ThrowInvalidOperationException_DerivedTypeIsAlreadySpecified(BaseType, derivedType);
+                }
+
+                if (typeDiscriminatorId is not null)
+                {
+                    if (!(_discriminatorIdtoType ??= new()).TryAdd(typeDiscriminatorId, derivedJsonTypeInfo))
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_TypeDicriminatorIdIsAlreadySpecified(BaseType, typeDiscriminatorId);
+                    }
+                }
+
+                containsDerivedTypes = true;
+            }
+
+            if (!containsDerivedTypes)
+            {
+                ThrowHelper.ThrowInvalidOperationException_PolymorphicTypeConfigurationDoesNotSpecifyDerivedTypes(BaseType);
+            }
+
+            if (UsesTypeDiscriminators)
+            {
+                if (!baseConverter.CanHaveMetadata)
+                {
+                    ThrowHelper.ThrowNotSupportedException_BaseConverterDoesNotSupportMetadata(BaseType);
+                }
+
+                if (configuration.CustomTypeDiscriminatorPropertyName is string customPropertyName)
+                {
+                    JsonEncodedText jsonEncodedName = JsonEncodedText.Encode(customPropertyName, options.Encoder);
+
+                    // Check if the property name conflicts with other metadata property names
```

Comment:
Are we opening ourselves up to breaking user apps in the future if they use a custom metadata tag that's unused in `STJ` now but then used in future releases?

Reply:
I wouldn't think so, this particular code path is only exercised for types that have explicitly opted into polymorphism which is new functionality.